Store courses in sessions

ryan-conn opened the pull request:
## Description

Saves courses when generate schedules is clicked, and sets courses on loading the page or choosing a term.

## Rationale

I had to change the `courseCards` redux pretty substantially to make everything work, so if there are any changes you guys want to suggest then I'm open to it. The main thing that's a bit strange with it is that `replaceCourseCards` doesn't actually create an action, it just dispatches several different actions depending on the arguments. The reason for this is so that all the intermediate states can be shown instead of having to calculate everything and then updating state, but I can change it to do everything simultaneously.

I also changed the formatting of the `sessions/get_last_term` API to return empty json instead of an empty string if a last term doesn't exist since it's simpler, didn't require any frontend changes, and is consistent with how the `get_saved_courses` API works.

Last thing is that I fetch updated sections and update the saved course cards since the seats/available sections, etc might have changed since the schedule was generated, I tried to keep as much information as possible from the saved schedules though. It might be worthwhile to simplify how schedules are represented on the backend (instead of serializing the entire `courseCards` redux slice, only store certain parts to reduce storage requirements) because not all of the information is used, but for now I'm keeping it simple. 

## How to test

Do whatever you want with course cards, click "generate schedules", and refresh/change terms, etc., and the state of your course cards should be exactly the same.

## Related tasks

Related to #260, still need to save availabilities but courses are pretty complicated as it involves fetching/modifying data rather than direct serialization/deserialization so it's not happening in this PR

ryan-conn commented:
Should be ready for review now, tests have been created

I also noticed that github actions wasn't running tests from the `user_sessions` folder, so I removed `testpaths` from our `pytest.ini` because it can find all tests ending in `test[s].py` by itself, I opted to do that instead of adding the `user_sessions/tests` folder to prevent this from happening again if we add another folder in the future

firejake308 commented on autoscheduler/frontend/src/redux/actions/courseCards.ts:
```diff
@@ -212,3 +234,71 @@ export function updateCourseCard(index: number, courseCard: CourseCardOptions, t
+    const newCourseCards = cards.map((card) => (updateSectionsForCourseCard(card, term)));
+
+    // clear all course cards
+    new Promise((resolve) => { dispatch(clearCourseCards()); resolve(); }).then(async () => {
```
Do you need to call `resolve` manually here?

ryan-conn replied:
Yes, if you don't do it manually then it doesn't wait for the dispatch to complete

gannonprudhomme replied:
In that case, I would mention that in a comment as it's not very intuitive to me

ryan-conn replied:
I changed the name of the array to `courseCardPromises` after Adel's comment, do you still think the comment is necessary? Also since I refactored this code to call `.then()` on each of them I thought it would be clearer

gannonprudhomme replied:
I think really what isn't clear to me is that you have to call `resolve()` to wait for the `dispatch(clearCourseCards())` to finish (still don't understand that tbh) - the `courseCardPromises` itself doesn't confuse me as much

ryan-conn replied:
Actually I don't think it's needed, I've been testing without it and it seems to work fine. Since the action is synchronous the function should have to wait until after it finishes to continue

firejake308 commented on autoscheduler/frontend/src/redux/actions/courseCards.ts:
```diff
@@ -212,3 +234,71 @@ export function updateCourseCard(index: number, courseCard: CourseCardOptions, t
+        // update information from all course cards and add them in order
+        for (let i = 0; i < newCourseCards.length; i++) {
+          // eslint-disable-next-line no-await-in-loop
+          const newCourseCard = await newCourseCards[i];
```
Why do we need an `await` here? Array access isn't asynchronous, is it?

ryan-conn replied:
We want the courses to be in the same order as they're stored in when they're recreated, so here it has to wait for the previous card to be generated before dispatching to add the next card

firejake308 commented on autoscheduler/frontend/src/redux/reducers/courseCards.ts:
```diff
@@ -61,6 +63,8 @@ export default function courseCards(
         [action.index]: { ...state[action.index], ...action.courseCard },
       };
+    case CLEAR_COURSE_CARDS:
+      return { numCardsCreated: 0 };
```
This might need a `...state`, let me see if I can check it

firejake308 replied:
Ok nvm it's fine

ryan-conn commented:
> As far as functionality is concerned, does this only save when you press Generate Schedules?

Yeah for now that's the only time they're sent, but during the meeting I was planning on talking about what other situations we'll want to save courses

gannonprudhomme commented on autoscheduler/frontend/src/tests/ui/CourseSelectColumn.test.tsx:
```diff
@@ -19,16 +23,20 @@ function ignoreInvisible(content: string, element: HTMLElement, query: string |
           <CourseSelectColumn />
         </Provider>,
       );
+      await new Promise(setImmediate);
```
What does this do? And why do we need it

ryan-conn replied:
This waits for all pending promises to be resolved, since we have to wait for all of the actions called by `replaceCourseCards` to finish if this isn't here then the lines after the dispatch will run before it's completed

ryan-conn replied:
Actually I don't think it's needed, I've been testing without it and it seems to work fine. Since the action is synchronous it should have to wait until after it finishes to continue

gannonprudhomme commented on autoscheduler/user_sessions/tests/courses_api_tests.py:
```diff
@@ -0,0 +1,108 @@
+    def test_save_courses_correctly_parses_course_data_json(self):
+        """ Tests that /sessions/save_courses can parse a complicated json object.
+            Note that this doesn't test anything related to the actual formatting:
+            it should simply trust that the frontend gives it data that can it can
```
```suggestion
            it should simply trust that the frontend gives it data that it can
```

gannonprudhomme commented on autoscheduler/user_sessions/tests/courses_api_tests.py:
```diff
@@ -0,0 +1,108 @@
+        response = self.client.get(f'/sessions/get_saved_courses?term={term}')
+
+        # Assert
+        self.assertNotEqual(response.status_code, 200)
```
Do you think asserting that it's not 200 is better than asserting that it is 400?

ryan-conn replied:
I might as well do it that way since it's more specific, when I made this test though I was thinking just to test that it doesn't give a successful response

gannonprudhomme commented on autoscheduler/frontend/src/components/SchedulingPage/ConfigureCard/ConfigureCard.tsx:
```diff
@@ -83,6 +84,29 @@ const ConfigureCard: React.FC = () => {
+    for (let i = 0; i < courseCards.numCardsCreated; i++) {
+      const course = courseCards[i];
+      if (course) {
+        const sections = course.sections.filter(({ selected }) => selected).map((section) => (
+          section.section.id
+        ));
```
```suggestion
        const sections = course.sections.filter(({ selected }) => selected).map((sectionSel) => (
          sectionSel.section.id
        ));
```

nit: Frankly I always get confused by `SectionSelected`'s, so I would prefer specifying that its not just a normal `Section` object (the `section.section` looks pretty weird too lol)

ryan-conn replied:
I refactored this so it's no longer relevant

gannonprudhomme commented on autoscheduler/user_sessions/views.py:
```diff
@@ -1,22 +1,61 @@
+@api_view(['PUT'])
+@parser_classes([JSONParser])
+def save_courses(request):
+    """ API endpoint that saves course cards for user's current term. """
```
I think it would be worth mentioning that this/both route(s) can store any format of data, since really the only thing related to courses is that it saves the data with the `courses` key in the session.

ryan-conn replied:
Yeah I put a comment about it in a test instead of the actual API for some reason, will change

gannonprudhomme commented on autoscheduler/frontend/src/tests/redux/CourseCards.test.ts:
```diff
@@ -274,4 +284,109 @@ describe('Course Cards Redux', () => {
     });
   });
+
+  describe('deletes all course cards and resets numCardsCreated', () => {
+    test('when clearCourseCards is dispatched', () => {
```
```suggestion
  describe('clearCourseCards', () => {
    test('deletes all course cards and resets numCardsCreated, () => {
```

So while this does sort of invert our typical test structure, I think I prefer having like the "category" as the out-most `describe` as it groups all the test for that action under the same block (i.e. it specifies the system under test more directly). Also similar to how the first word/phase after `test_` in our Python tests are the system under test. Same thing with the rests of the tests

gannonprudhomme commented on autoscheduler/frontend/src/redux/actions/courseCards.ts:
```diff
@@ -153,6 +171,33 @@ function sortSections(sections: SectionSelected[]): SectionSelected[] {
+ * @param courseCard course to get sections for
+ * @param term term to fetch sections for
+ */
+async function createCourseCardFrom(
```
I would change the name of this to mention that it's doing a fetch (e.g. add a `fetch` or `retrieve` to the name)

gannonprudhomme commented on autoscheduler/frontend/src/tests/redux/CourseCards.test.ts:
```diff
@@ -274,4 +284,109 @@ describe('Course Cards Redux', () => {
+    });
+  });
+
+  describe('replaces all course cards and keeps attributes not related to sections', () => {
```
What does "keep attributes" mean here?

ryan-conn replied:
This should probably say "properties", I mean all of the data for the card

gannonprudhomme commented on autoscheduler/frontend/src/redux/actions/courseCards.ts:
```diff
@@ -153,6 +171,33 @@ function sortSections(sections: SectionSelected[]): SectionSelected[] {
+  courseCard: CourseCardOptions,
+  term: string,
+): Promise<CourseCardOptions> {
+  // const { course } = courseCard;
```
Remove this commented out line

gannonprudhomme commented on autoscheduler/frontend/src/redux/actions/courseCards.ts:
```diff
@@ -163,31 +208,12 @@ function updateCourseCardAsync(
+    createCourseCardFrom(courseCard, term).then((updatedCourseCard) => {
+      dispatch(updateCourseCardSync(index, updatedCourseCard));
+    }).catch((error) => {
+      // eslint-disable-next-line no-console
+      console.log(error);
```
I'm fine leaving this for now, but tbh we probably _shouldn't_ be outputting to the console - at least not in production. We can always add replace this with actual logging that would send this error somewhere we could view though (i.e. for crash analytics) / use a logging library so when we're debugging we can have it print to the console, but hide it when in prod.

ryan-conn replied:
If you look at the diff I just moved this into a function so this console log was there before, I agree but I didn't want to change what was already there. Should I just remove it?

firejake308 replied:
I'm thinking, yeah, we can remove it for now. I only put it in there as like a "just in case" sorta thing, but we can figure out a better crash-handling system later

ryan-conn commented:
> Are the selected sections intentionally not being saved / reloaded? It remembers the customization level we had selected, but not the actual sections I selected.
This is unintentional, when I serialized course data I forgot to change the way I calculate what sections are checked (instead of an array it was still expecting `CourseCardOptions`) and didn't update the tests.

To fix this in the future, I made an interface called `SerializedCourseCard` that is used when creating the serialized data and when creating course cards from it so they should be guaranteed to have the same data now

> Also, do we want to save whether a card was collapsed/expanded?

We can do this in a future PR, but since it would involve moving expanded/collapsed into redux instead of the `CourseSelectCard` state I don't want to mess with it now

ryan-conn commented:
I addressed all your comments that have a 👍 which is most of them, so this should be ready for another round of review

firejake308 commented on autoscheduler/frontend/src/components/SchedulingPage/CourseSelectColumn/CourseSelectColumn.tsx:
```diff
@@ -17,6 +17,22 @@ const CourseSelectColumn: React.FC = () => {
+    }).catch(() => {
+      // If unable to parse saved course cards or request fails, go back to default
+      // of one empty course card
+      dispatch(replaceCourseCards([], term));
```
Is there a difference between this and `clearCourseCards()`?

ryan-conn replied:
`replaceCourseCards` with an empty array creates the initial empty card, now that I look back at this I think I could just make the initial state a single course card again, and then if there are saved course cards it should be able to correctly update the initial blank one and avoid ambiguity like this

firejake308 commented on autoscheduler/frontend/src/redux/actions/courseCards.ts:
```diff
@@ -212,3 +238,66 @@ export function updateCourseCard(index: number, courseCard: CourseCardOptions, t
+    const selectedSections = new Set(courseCard.sections);
+
+    newCourseCard.sections.forEach((section) => {
+      // eslint-disable-next-line no-param-reassign
```
Instead of disabling the rule, we should be able to just set the `props` option to false in the config, since we're reassigning a property of the param, not the param itself

firejake308 commented on autoscheduler/frontend/src/redux/actions/courseCards.ts:
```diff
@@ -212,3 +238,66 @@ export function updateCourseCard(index: number, courseCard: CourseCardOptions, t
+/**
+ * Creates an updated course card from courseCard by making an API call for updated
+ * section information
+ * @param courseCard Course card to update
```
Can you specify here that `courseCard` contains data saved from the last visit?

firejake308 commented on autoscheduler/frontend/src/redux/actions/courseCards.ts:
```diff
@@ -212,3 +238,66 @@ export function updateCourseCard(index: number, courseCard: CourseCardOptions, t
+  // for each one asynchronously and update them with data from courseCards
+  return (dispatch): void => {
+    // if courseCards is improperly formatted or no cards are saved, create default
+    if (!Array.isArray(courseCards) || courseCards.length === 0) {
```
What else would `courseCards` be besides an array? Is this different from null-checking?

ryan-conn replied:
The default response is `{}` instead of `[]` since I didn't want to change the expected format, I could just change it to `[]` instead to make this easier to understand

firejake308 commented on autoscheduler/frontend/src/redux/actions/courseCards.ts:
```diff
@@ -212,3 +238,66 @@ export function updateCourseCard(index: number, courseCard: CourseCardOptions, t
+    }
+
+    // create promise for each updated course card
+    const courseCardPromises = courseCards.map((card) => (updateSectionsForCourseCard(card, term)));
```
I feel like you have one extra set of parentheses here

firejake308 commented on autoscheduler/frontend/src/tests/redux/CourseCards.test.ts:
```diff
@@ -274,4 +284,94 @@ describe('Course Cards Redux', () => {
+
+  describe('clearCourseCards', () => {
+    test('deletes all course cards and resets numCardsCreated', () => {
+      // arrange
```
I mean, there was no data to clear in this case, so this test would've passed even without dispatching any actions. I'd probably add a card during the arrange phase just to make sure that clearing it actually does something.

firejake308 commented on autoscheduler/frontend/src/tests/redux/CourseCards.test.ts:
```diff
@@ -274,4 +284,94 @@ describe('Course Cards Redux', () => {
+      expect(store.getState().courseCards).toMatchObject(expectedCourseCards);
+    });
+
+    test('adds non-existing sections for course cards', async () => {
```
```suggestion
    test('adds non-pre-existing sections for course cards', async () => {
```

firejake308 commented on autoscheduler/frontend/src/tests/redux/CourseCards.test.ts:
```diff
@@ -274,4 +284,94 @@ describe('Course Cards Redux', () => {
+      const section = 830262;
+      fetchMock.mockImplementationOnce(testFetch);
+
+      const courseCards = [
```
This is the serialized version, right?

ryan-conn replied:
Yeah, I should declare this as `SerializedCourseCardOptions[]` to make it more clear

ryan-conn commented:
Fixed the issue where cards don't get replaced when browser navigation is used (thanks for the suggestion, I forgot you could add cleanup to `useEffect`), and everything else should be addressed now, I changed the default back to 1 course card again since `addCourseCard` can overwrite existing cards now

firejake308 commented on autoscheduler/frontend/src/tests/redux/CourseCards.test.ts:
```diff
@@ -274,4 +286,108 @@ describe('Course Cards Redux', () => {
+      await new Promise(setImmediate);
+
+      // assert
+      // testFetch with a MATH course has 1 section, which should be added
```
Do y'all think we need to add one more test to make sure that the `replaceCourseCards` action works if the `fetch`es return out of order? So, for example, if the `/api/sections` call returned card #1 before card #0, they stay in order? Intuitively, it feels like something that could go wrong if someone messed with it, but Ryan's implementation right now is solid enough to where it feels like common sense, so I'm not sure.

ryan-conn replied:
I added one just so that we have more thorough tests, and just in case we change the implementation later we can still know that this works

ryan-conn commented:
> Just had one idea for a potential test, but I don't have many more suggestions besides some of the ones from before that are still unresolved.

I added the test you suggested and fixed a couple of suggestions that I missed the first time around, everything should be addressed now
